### PR TITLE
add tonemapping LUT bindings for sprite and mesh2d pipelines

### LINKED ISSUE #13118: Crash when using Camera2DBundle without HDR + TonyMcMapFace tonemapping
## Bevy version

0.13.2, `main` branch (22d605c8)

## \[Optional\] Relevant system information

If you cannot get Bevy to build or run on your machine, please include:

- Latest rust version
- Windows 11

If your bug is rendering-related, copy the adapter info that appears when you run Bevy.

```ignore
AdapterInfo { 
  name: "NVIDIA GeForce RTX 3080", vendor: 4318, device: 8710, 
  device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "551.86", backend: Vulkan 
}
```

You should also consider testing the examples of our upstream dependencies to help isolate any setup-specific issue:

- [`wgpu`](https://github.com/gfx-rs/wgpu) for rendering problems
- [`winit`](https://github.com/rust-windowing/winit) for input and window management
- [`gilrs`](https://docs.rs/gilrs/latest/gilrs/) for gamepad inputs

## What you did

TL;DR:

- Spawn a `Camera2DBundle`
  - Set `HDR` = false
  - Set `tonemapping` = `Tonemapping::TonyMcMapFace` 
- Run Bevy

Minimal code snippet:

```rust
use bevy::{
    core_pipeline::tonemapping::Tonemapping,
    prelude::*
};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn((
        Camera2dBundle {
            camera: Camera {
                hdr: false,
                ..default()
            },
            tonemapping: Tonemapping::TonyMcMapface,
            ..default()
        },
    ));
}
```

```toml
[package]
name = "bevy_crash_demo"
version = "0.1.0"
edition = "2021"

[dependencies]
bevy = "0.13.2"

## What's maybe interesting is that this seems to be relevant to the feature `bevy_ui`:
## no crash:
# bevy = { version = "0.13.2", default-features = false, features = ["bevy_core_pipeline",  "tonemapping_luts", "bevy_winit"]} 

## crash:
# bevy = { version = "0.13.2", default-features = false, features = ["bevy_core_pipeline",  "tonemapping_luts", "bevy_winit", "bevy_ui"]} # crash
```

Describe how you arrived at the problem. If you can, consider providing a code snippet or link.

## What went wrong

```
2024-04-27T14:40:26.471379Z ERROR log: Device::create_render_pipeline error: Error matching ShaderStages(FRAGMENT) shader requirements against the pipeline    
2024-04-27T14:40:26.471813Z ERROR log: Device::create_render_pipeline error: Error matching ShaderStages(FRAGMENT) shader requirements against the pipeline
2024-04-27T14:40:26.471844Z ERROR log: Handling wgpu errors as fatal by default
thread 'Async Compute Task Pool (1)' panicked at C:\Users\BuuGh\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wgpu-0.19.4\src\backend\wgpu_core.rs:3006:5:
wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `sprite_pipeline`
    Error matching ShaderStages(FRAGMENT) shader requirements against the pipeline
    Shader global ResourceBinding { group: 0, binding: 18 } is not available in the pipeline layout
    Binding is missing from the pipeline layout
```

Backtrace: https://github.com/bevyengine/bevy/issues/13118#issuecomment-2081120194

## Additional information

Other information that can be used to further reproduce or isolate the problem.
This commonly includes:

- Things like `Reinhard` for tonemapping works fine, it seems specific to `TonyMcMapFace`.

### PR DESCRIPTION
**Pull request by arcashka**
Fixes #13118
If you use `Sprite` or `Mesh2d` and create `Camera` with
* hdr=false
* any tonemapper

You would get
```
wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `sprite_pipeline`
    Error matching ShaderStages(FRAGMENT) shader requirements against the pipeline
    Shader global ResourceBinding { group: 0, binding: 19 } is not available in the pipeline layout
    Binding is missing from the pipeline layout
```
Because of missing tonemapping LUT bindings 

## Solution
Add missing bindings for tonemapping LUT's to `SpritePipeline` & `Mesh2dPipeline`

## Testing
I checked that
* `tonemapping`
* `color_grading`
* `sprite_animations`
* `2d_shapes`
* `meshlet`
* `deferred_rendering`
examples are still working

2d cases I checked with this code:
```
use bevy::{
    color::palettes::css::PURPLE, core_pipeline::tonemapping::Tonemapping, prelude::*,
    sprite::MaterialMesh2dBundle,
};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .add_systems(Update, toggle_tonemapping_method)
        .run();
}

fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<ColorMaterial>>,
    asset_server: Res<AssetServer>,
) {
    commands.spawn(Camera2dBundle {
        camera: Camera {
            hdr: false,
            ..default()
        },
        tonemapping: Tonemapping::BlenderFilmic,
        ..default()
    });
    commands.spawn(MaterialMesh2dBundle {
        mesh: meshes.add(Rectangle::default()).into(),
        transform: Transform::default().with_scale(Vec3::splat(128.)),
        material: materials.add(Color::from(PURPLE)),
        ..default()
    });

    commands.spawn(SpriteBundle {
        texture: asset_server.load("asd.png"),
        ..default()
    });
}

fn toggle_tonemapping_method(
    keys: Res<ButtonInput<KeyCode>>,
    mut tonemapping: Query<&mut Tonemapping>,
) {
    let mut method = tonemapping.single_mut();

    if keys.just_pressed(KeyCode::Digit1) {
        *method = Tonemapping::None;
    } else if keys.just_pressed(KeyCode::Digit2) {
        *method = Tonemapping::Reinhard;
    } else if keys.just_pressed(KeyCode::Digit3) {
        *method = Tonemapping::ReinhardLuminance;
    } else if keys.just_pressed(KeyCode::Digit4) {
        *method = Tonemapping::AcesFitted;
    } else if keys.just_pressed(KeyCode::Digit5) {
        *method = Tonemapping::AgX;
    } else if keys.just_pressed(KeyCode::Digit6) {
        *method = Tonemapping::SomewhatBoringDisplayTransform;
    } else if keys.just_pressed(KeyCode::Digit7) {
        *method = Tonemapping::TonyMcMapface;
    } else if keys.just_pressed(KeyCode::Digit8) {
        *method = Tonemapping::BlenderFilmic;
    }
}
```
---

## Changelog
Fix the bug which led to the crash when user uses any tonemapper without hdr for rendering sprites and 2d meshes.

### CODE REVIEW

**Comment by arcashka** [comment]
Rebased due to the conflcts

File: crates/bevy_sprite/src/mesh2d/mesh.rs [hunk 1]
```diff
@@ -263,14 +267,22 @@ impl FromWorld for Mesh2dPipeline {
         )> = SystemState::new(world);
         let (render_device, render_queue, default_sampler) = system_state.get_mut(world);
         let render_device = render_device.into_inner();
+        let tonemapping_lut_entries = get_lut_bind_group_layout_entries();
         let view_layout = render_device.create_bind_group_layout(
             "mesh2d_view_layout",
-            &BindGroupLayoutEntries::sequential(
+            &BindGroupLayoutEntries::with_indices(
```
Comment (jnhyatt) [inline crates/bevy_sprite/src/mesh2d/mesh.rs:273]: Why not keep this sequential?

File: crates/bevy_sprite/src/mesh2d/mesh.rs [hunk 2]
```diff
@@ -584,29 +604,42 @@ pub struct Mesh2dViewBindGroup {
...
+        return;
+    };
+
+    for (entity, tonemapping) in &views {
+        let lut_bindings =
+            get_lut_bindings(&images, &tonemapping_luts, tonemapping, &fallback_image);
+        let view_bind_group = render_device.create_bind_group(
+            "mesh2d_view_bind_group",
+            &mesh2d_pipeline.view_layout,
+            &BindGroupEntries::with_indices((
```
Comment (jnhyatt) [inline crates/bevy_sprite/src/mesh2d/mesh.rs:632]: Same here, any reason not to use sequential?

File: crates/bevy_sprite/src/render/mod.rs [hunk 3]
```diff
@@ -528,13 +554,46 @@ pub fn queue_sprites(
 }
 
 #[allow(clippy::too_many_arguments)]
-pub fn prepare_sprites(
```
Comment (jnhyatt) [inline crates/bevy_sprite/src/render/mod.rs:531]: This diff is pretty gnarly but it looks like it's just a (unrelated?) refactor of this system into two, right?

**Review by jnhyatt — APPROVED** [review]
Generally looks good, just looks like half the diff is miscellaneous cleanup? Not a problem, just want to make sure I'm not missing something. Couple of unimportant style things. Otherwise lgtm

File: crates/bevy_sprite/src/render/mod.rs (hunk 3 above)
Comment (arcashka) [reply crates/bevy_sprite/src/render/mod.rs:531]: It's not unrelated. These systems are split into two because one of them does per-sprite setup, `prepare_sprite_image_bind_groups`, and the other one does per-view setup, `prepare_sprite_view_bind_groups`.

Before this PR, it was done in one system because earlier view bind groups were always the same, even for different views. Now, since different views (cameras) can have different tonemapping settings, we need to be able to provide different LUTs for different views.
It's pretty close to what is currently done for
* 3d meshes: https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/view/mod.rs#L788
* 2d meshes: https://github.com/bevyengine/bevy/blob/main/crates/bevy_sprite/src/mesh2d/mesh.rs#L587

File: crates/bevy_sprite/src/mesh2d/mesh.rs (hunk 1 above)
Comment (arcashka) [reply crates/bevy_sprite/src/mesh2d/mesh.rs:273]: From what I read in the Bevy Discord, the common consensus is that it's better to use explicit indices instead of sequential

**Comment by arcashka** [comment]
> Generally looks good, just looks like half the diff is miscellaneous cleanup? Not a problem, just want to make sure I'm not missing something. Couple of unimportant style things. Otherwise lgtm

@jnhyatt thank you for the review :heart: 
I rechecked the changes, most of them are not unrelated.
There are bunch of shader reorganization changes, but they are required. Because lut bindings are defined like this
```
@group(0) @binding(#TONEMAPPING_LUT_TEXTURE_BINDING_INDEX) var dt_lut_texture: texture_3d<f32>;
@group(0) @binding(#TONEMAPPING_LUT_SAMPLER_BINDING_INDEX) var dt_lut_sampler: sampler;
```
tonemapping_shared.wgsl can not be imported unless `TONEMAPPING_LUT_TEXTURE_BINDING_INDEX` and `TONEMAPPING_LUT_SAMPLER_BINDING_INDEX` are defined.
So I had to move `pow_safe` from tonemapping_shared because it's being used outside of tonemapping in some cases.

File: crates/bevy_core_pipeline/src/tonemapping/lut_bindings.wgsl
[the remark this answers is not in the document]
Comment (kristoff3r) [reply crates/bevy_core_pipeline/src/tonemapping/lut_bindings.wgsl:1]: Maybe link that issue in the file or where it is imported? I couldn't figure out from skimming the issue if you feel like a cleaner solution is possible later.

**Review by kristoff3r — APPROVED** [review]
LGTM, also tested tonemapping example on wasm

**Comment by alice-i-cecile** [comment]
@arcashka once merge conflicts are resolve I'll merge this in :)

**Comment by arcashka** [comment]
> Maybe link that issue in the file or where it is imported? I couldn't figure out from skimming the issue if you feel like a cleaner solution is possible later.

I thought about it but decided against it because having bindings in a separate module is pretty common within the Bevy source code.
image

I don't think having lut bindings in the separate file is something which can cause questions/needs explaining. It's a standard practive and by accident also a way to bypass this weird naga-oil issue.

**Comment by arcashka** [comment]
> @arcashka once merge conflicts are resolve I'll merge this in :)

Rebased and resolved merge conflicts :)
